File: java/src/main/java/gherkin/ParserException.java
package gherkin;

import gherkin.ast.Location;

import java.util.List;

public class ParserException extends RuntimeException {
    public Location location;

    protected ParserException(String message) {
        super(message);
    }

    protected ParserException(String message, Location location) {
        super(GetMessage(message, location));
        if (location == null) throw new NullPointerException("location");

        this.location = location;
    }

    private static String GetMessage(String message, Location location) {
        if (location == null) throw new NullPointerException("location");

        return String.format("(%s:%s): %s", location.line, location.column, message);
    }

    public static class AstBuilderException extends ParserException {
        public AstBuilderException(String message) {
            super(message);
            if (message == null) throw new NullPointerException("message");
        }

        public AstBuilderException(String message, Location location) {
            super(message, location);
        }
    }

    public static class UnexpectedTokenException extends ParserException {
        public String StateComment;

        public final Token ReceivedToken;
        public final List<String> ExpectedTokenTypes;

        public UnexpectedTokenException(Token receivedToken, List<String> expectedTokenTypes, String stateComment) {
            super(GetMessage(receivedToken, expectedTokenTypes), receivedToken.location);
            ReceivedToken = receivedToken;
            ExpectedTokenTypes = expectedTokenTypes;
            StateComment = stateComment;
        }

        private static String GetMessage(Token receivedToken, List<String> expectedTokenTypes) {
            if (receivedToken == null) throw new NullPointerException("receivedToken");
            if (expectedTokenTypes == null) throw new NullPointerException("expectedTokenTypes");

            return String.format("expected: %s, got '%s'",
                    StringUtils.join(", ", expectedTokenTypes),
                    receivedToken.getTokenValue().trim());
        }
    }

    public static class UnexpectedEOFException extends ParserException {
        public final String StateComment;
        public final List<String> ExpectedTokenTypes;

        public UnexpectedEOFException(Token receivedToken, List<String> expectedTokenTypes, String stateComment) {
            super(GetMessage(expectedTokenTypes), receivedToken.location);
            ExpectedTokenTypes = expectedTokenTypes;
            StateComment = stateComment;
        }

        private static String GetMessage(List<String> expectedTokenTypes) {
            if (expectedTokenTypes == null) throw new NullPointerException("expectedTokenTypes");

            return String.format("unexpected end of file, expected: %s",
                    StringUtils.join(", ", expectedTokenTypes));
        }
    }

    public static class CompositeParserException extends ParserException {
        public final List<ParserException> Errors;

        public CompositeParserException(List<ParserException> errors) {
            super(GetMessage(errors));
            if (errors == null) throw new NullPointerException("errors");

            Errors = errors;
        }

        private static String GetMessage(List<ParserException> errors) {
            if (errors == null) throw new NullPointerException("errors");

            StringUtils.ToString<ParserException> exceptionToString = new StringUtils.ToString<ParserException>() {
                @Override
                public String toString(ParserException e) {
                    return e.getMessage();
                }
            };
            return "Parser errors: \n" + StringUtils.join(exceptionToString, "\n", errors);
        }
    }
}

File: Gherkin/ParserException.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Gherkin.Ast;

namespace Gherkin
{
    public abstract class ParserException : Exception
    {
        public Location Location { get; private set; }

        protected ParserException(string message) : base(message)
        {
        }

        protected ParserException(string message, Location location) : base(GetMessage(message, location))
        {
            if (location == null) throw new ArgumentNullException("location");

            Location = location;
        }

        private static string GetMessage(string message, Location location)
        {
            if (location == null) throw new ArgumentNullException("location");

            return string.Format("({0}:{1}): {2}", location.Line, location.Column, message);
        }
    }

    public class AstBuilderException : ParserException
    {
        public AstBuilderException(string message) : base(message)
        {
            if (message == null) throw new ArgumentNullException("message");
        }

        public AstBuilderException(string message, Location location) : base(message, location)
        {
            if (message == null) throw new ArgumentNullException("message");
            if (location == null) throw new ArgumentNullException("location");
        }
    }

    public class NoSuchLanguageException : ParserException
    {
        public NoSuchLanguageException(string language, Location location) :
            base("Language not supported: " + language, location)
        {
        }
    }
    
    public class UnexpectedTokenException : ParserException
    {
        public string StateComment { get; private set; }

        public Token ReceivedToken { get; private set; }
        public string[] ExpectedTokenTypes { get; private set; }

        public UnexpectedTokenException(Token receivedToken, string[] expectedTokenTypes, string stateComment)
            : base(GetMessage(receivedToken, expectedTokenTypes), receivedToken.Location)
        {
            if (receivedToken == null) throw new ArgumentNullException("receivedToken");
            if (expectedTokenTypes == null) throw new ArgumentNullException("expectedTokenTypes");

            ReceivedToken = receivedToken;
            ExpectedTokenTypes = expectedTokenTypes;
            StateComment = stateComment;
        }

        private static string GetMessage(Token receivedToken, string[] expectedTokenTypes)
        {
            if (receivedToken == null) throw new ArgumentNullException("receivedToken");
            if (expectedTokenTypes == null) throw new ArgumentNullException("expectedTokenTypes");

            return string.Format("expected: {0}, got '{1}'",
                string.Join(", ", expectedTokenTypes),
                receivedToken.GetTokenValue().Trim());
        }
    }

    public class UnexpectedEOFException : ParserException
    {
        public string StateComment { get; private set; }
        public string[] ExpectedTokenTypes { get; private set; }
        public UnexpectedEOFException(Token receivedToken, string[] expectedTokenTypes, string stateComment)
            : base(GetMessage(expectedTokenTypes), receivedToken.Location)
        {
            if (expectedTokenTypes == null) throw new ArgumentNullException("expectedTokenTypes");

            ExpectedTokenTypes = expectedTokenTypes;
            StateComment = stateComment;
        }

        private static string GetMessage(string[] expectedTokenTypes)
        {
            if (expectedTokenTypes == null) throw new ArgumentNullException("expectedTokenTypes");

            return string.Format("unexpected end of file, expected: {0}",
                string.Join(", ", expectedTokenTypes));
        }
    }

    public class CompositeParserException : ParserException
    {
        public IEnumerable<ParserException> Errors { get; private set; }

        public CompositeParserException(ParserException[] errors)
            : base(GetMessage(errors))
        {
            if (errors == null) throw new ArgumentNullException("errors");

            Errors = errors;
        }

        private static string GetMessage(ParserException[] errors)
        {
            if (errors == null) throw new ArgumentNullException("errors");

            return "Parser errors:" + Environment.NewLine + string.Join(Environment.NewLine, errors.Select(e => e.Message));
        }
    }
}
